plugins/org.polarsys.capella.scenario.editor.dsl/src/org/polarsys/capella/scenario/editor/dsl/validation/TextualScenarioValidator.xtend

```diff
@@ -373,42 +421,58 @@ class TextualScenarioValidator extends AbstractTextualScenarioValidator {
 		}
 
 	}
-
+	
 	/*
 	 * Check that each withExecution message is closed by deactivation (on the proper target)
 	 */
 	@Check
-	def checkWithExecutionHasDeactivate(Model model) {
+	def checkWithExecutionHasDeactivateModel(Model model) {
+		var messageWithExecutionTargets = checkWithExecutionHasDeactivate(model)
+		for (String target : messageWithExecutionTargets.keySet) {
+			error(
+				'Deactivation keyword expected for a withExecution message',
+				TextualScenarioPackage.Literals.MODEL__ELEMENTS,
+				messageWithExecutionTargets.get(target)
+			)
+		}
+	}
+	
+	@Check
+	def checkWithExecutionHasDeactivateBlock(Block block) {
+		var messageWithExecutionTargets = checkWithExecutionHasDeactivate(block)
+		for (String target : messageWithExecutionTargets.keySet) {
+			error(
+				'Deactivation keyword expected for a withExecution message',
+				TextualScenarioPackage.Literals.BLOCK__BLOCK_ELEMENTS,
+				messageWithExecutionTargets.get(target)
+			)
+		}
+	}
+		
+	def checkWithExecutionHasDeactivate(EObject model) {
 		// keep a list with the target of the messages that contains the withExecution keyword
 		// keep also a list with the index on which withExecution message is found, to know on which line to show an error
-		var messageWithExecutionTargets = newLinkedList
-		var messageWithExecutionTargetsIndex = newLinkedList
+		var messageWithExecutionTargets = new HashMap
 		var index = 0
-		for (obj : model.elements) {
+		var elements = TextualScenarioHelper.getElements(model)
+		for (obj : elements) {
 			if (obj instanceof SequenceMessage && (obj as SequenceMessage).execution !== null) {
 				// add the SequenceMessage with execution to a list
-				messageWithExecutionTargets.add((obj as SequenceMessage).target)
-				messageWithExecutionTargetsIndex.add(index)
+				messageWithExecutionTargets.put((obj as SequenceMessage).target, index)

```

Comment:
If you have 2 messages with the same target (but different source) it will override the message, as the key in map is the target.

Valid:
A -> B withExecution: M1
C -> B withExecution: M2
deactivate B
deactivate B